File: src/Cpp/09_Listy_wprowadzenie/Zad14.cpp
#include <algorithm>
#include <cassert>
#include <numeric>
#include <vector>

/*
Dla otrzymanej listy składającej się z nieparzystej liczby
elementów znajdź element bez pary. Przez pary elementów rozumiemy
tutaj elementy o równej wartości.
*/

// Zlozonosc czasowa O(n)
// Zlozonosc pamieciowa O(n)
int elementBezParyV1(std::vector<int> &lista) {
  std::vector<int> pom;

  for (auto liczba : lista) {
    auto i = find(pom.begin(), pom.end(), liczba);
    if (i != pom.end())
      pom.erase(i);
    else
      pom.push_back(liczba);
  }

  return pom.front();
}

// Zlozonosc czasowa O(n)
// Zlozonosc pamieciowa O(n)
int elementBezParyV2(std::vector<int> &lista) {
  int wynik = 0;

  for (auto liczba : lista)
    wynik ^= liczba;

  return wynik;
}

void test1() {
  std::vector<int> lista{1, 3, 1, 7, 3, 1, 1};
  int wynik = 7;
  assert(elementBezParyV1(lista) == wynik);
}

void test2() {
  std::vector<int> lista{1, 3, 1, 7, 3, 1, 1};
  int wynik = 7;
  assert(elementBezParyV2(lista) == wynik);
}

int main() {

  test1();
  test2();

  return 0;
}

File: src/Python/09_Listy_wprowadzenie/Zad14.py
"""
Dla otrzymanej listy składającej się z nieparzystej liczby 
elementów znajdź element bez pary. Przez pary elementów rozumiemy 
tutaj elementy o równej wartości. 
"""

def element_bez_pary(lista):
    for i in lista:
        if lista.count(i) % 2 == 1:
            return i

def test_element_bez_pary():
    assert element_bez_pary([1, 3, 1, 7, 3, 1, 1]) == 7
    assert element_bez_pary([1, 1, 2, 2, 3, 4, 4, 5, 5]) == 3
    assert element_bez_pary([1, 1, 2, 2, 3, 3, 4, 4, 4, 4, 2]) == 2

if __name__ == "__main__":
    
    test_element_bez_pary()